src/components/Table/useSortState.ts

```diff
@@ -19,35 +19,35 @@ export function useSortState<R extends Kinded, S>(
   // If we're server-side sorting, use the caller's `sorting.value` prop to initialize our internal
   // `useState`. After this, we ignore `sorting.value` because we assume it should match what our
   // `setSortState` just changed anyway (in response to the user sorting a column).
-  const [sortState, setSortState] = useState<SortState<S> | undefined>(() => {
+  const initialSortState: SortState<S> | undefined = useMemo(() => {
     if (sorting?.on === "client") {
       const { initial } = sorting;
-      if (initial) {
+      if (initial === undefined && "initial" in sorting) {
+        // if explicitly set to `undefined`, then do not sort
+        return undefined;
+      } else if (initial) {
         const key = typeof initial[0] === "number" ? initial[0] : columns.indexOf(initial[0] as any);
         return [key as any as S, initial[1]];
       } else {
         // If no explicit sorting, assume 1st column ascending
         const firstSortableColumn = columns.findIndex((c) => c.clientSideSort !== false);
-        return [firstSortableColumn as any as S, "ASC"];
+        return [firstSortableColumn as any as S, ASC];
       }
     } else {
       return sorting?.value;
     }
-  });
+  }, [sorting, columns]);
+  const [sortState, setSortState] = useState<SortState<S> | undefined>(initialSortState);
 
-  // Make a custom setSortKey that is useState-like but contains the invert-if-same-column-clicked-twice logic.
+  // Make a custom setSortKey that is useState-like but contains the ASC->DESC->RESET logic.
   const setSortKey = useCallback(
     (clickedKey: S) => {
-      const [currentKey, currentDirection] = sortState || [];
-      const [newKey, newDirection] =
-        // If clickedKey === currentKey, then toggle direction
-        clickedKey === currentKey
-          ? [currentKey, currentDirection === ASC ? DESC : ASC]
-          : // Otherwise, use the new key, and default to ascending
-            [clickedKey, ASC];
-      setSortState([newKey, newDirection]);
+      const newState = deriveSortState(sortState, clickedKey, initialSortState);
+      setSortState(newState);
+
       if (sorting?.on === "server") {
-        sorting.onSort(newKey, newDirection);
+        const [initialKey, initialDirection] = newState ? newState : [undefined, undefined];
```

Comment:
Could this be `newState ?? [undefined, undefined]`? 

Could `initialKey` / `initialDirection` be called `newKey` / `newDirection`?

Reply:
Yes and Yes.